### PR TITLE
fix(dsp): discard the partial panadapter frame at a sequence gap, in both raw-IQ backends

### PR DESCRIPTION
**Pull request by on8st**
A transport sequence discontinuity could join pre- and post-discontinuity samples in one FFT window on HL2 and ANAN. The spectrum accumulators already had `reset()`, but the transports never notified the DSP to use it.

This change delivers a continuity notification before the affected datagram's IQ block and resets every affected receiver's partial spectrum window. Forward packet loss, accepted rewinds, duplicates, and ambiguous half-range jumps all invalidate that window. In-order delivery, including normal uint32 sequence wraparound, continues to accumulate normally.

Packet-loss accounting stays unchanged: HL2 counts forward missing packets, while ANAN retains its existing one-per-sequence-mismatch counter. HL2's `rxSequenceGap(lost)` carries zero on a backward/duplicate transition; its notification is about continuity independently of loss accounting. Audio buffers, WDSP state, display-rate clocks, and ANAN's magnitude smoothing are unchanged.

## Diagnostics

Both DSP stages expose `spectrumGapDiscards()`, incremented only when a nonempty FFT window is discarded. HL2 publishes `spectrumGapDiscards<n>` per receiver through `health`; this is an additive script-facing diagnostic. ANAN still has no health rows.

The discard count is monotonic for the DSP object's lifetime and survives sample-rate reconfiguration. It records prevention of a discontinuous FFT, not a corrupted frame displayed. Repeated discontinuities can prevent a complete window and leave the last trace displayed; measurement scripts should check frame liveness and counter deltas. The automation documentation describes these semantics and null values during receiver rebuilds.

## Coverage and validation

The registered `spectrum_sequence_gap_test` is socket-free: neither transport is started, no socket is created or bound, and there is no radio peer. ANAN's existing datagram handling is extracted behind a private friend seam so the test exercises the production parsing, per-DDC sequence tracking, notification ordering, and IQ delivery, as the existing Metis seam does.

Coverage includes:

- Both spectrum resets and both receive DSP stages, with exact bin equality against clean post-discontinuity input.
- Both wire ingest handlers through their receive DSPs: forward loss, rewind, duplicate, contiguous delivery, normal wraparound, loss across wraparound, and the ambiguous half-range jump.
- Gap-before-IQ ordering, unchanged loss counters, empty-window behavior, and the display-rate rolling accumulator.
- ANAN per-DDC isolation, wrong-port and malformed-datagram rejection, and DDC0-only link establishment.

Local validation on Apple Silicon, RelWithDebInfo, RADE enabled, against main `aea9f6dc4c62c6cc8ad84159e7050ea6a1e728f8`:

```sh
cmake --build build-codex-arm64 --target spectrum_sequence_gap_test hl2_spectrum_test anan_p2_protocol_test -j8
ctest --test-dir build-codex-arm64 -R '^(spectrum_sequence_gap_test|hl2_spectrum_test|anan_p2_protocol_test)$' --no-tests=error --output-on-failure
```

All three tests passed. After main advanced to `5c146c68b52581d7ccbf9113fe0b4f56b1bf270e` during hosted CI, the combined tree was rebuilt locally and all three tests passed again (tree `6d8b2a40ebb76cf41c00764edb4dbff57cee67d1`). ARM64 processor configuration and executable verified; no RNNoise x86 sources in the build graph. Engine-boundary, test-registration, frozen CI test-gate, and touchpoint-manifest checks passed (engine-boundary reports only tracked legacy warnings).

Mutation evidence: restoring the old forward-only notification guards on both transports produces 29 assertion failures; restoring the repair returns the test to green. The initial independent review also disabled both DSP reset handlers and caught four failures.

Limits: the test connects real transport and DSP objects itself; backend constructor wiring is source-reviewed, not exercised by this target. No GUI, real-radio, synthetic-peer, or TX session was run. Demo mode does not exercise HL2/ANAN ingest. No full-suite result is claimed. The new test is registered for full-suite lanes but is not added to the frozen per-PR test execution gate.

Architecture conformance: all four gates (engine boundary, capability records, command plane, touchpoint manifest) were run at base `aea9f6dc` and published head `a1ec80fa`; their per-file findings are identical. The gates were also rerun at latest main `5c146c68` and the tested combined tree using the updated checkers; all four again reported identical findings. Manifest regeneration on the latest combined tree produces no unstaged change.

### CODE REVIEW

File: src/core/backends/hl2/Hl2RxDsp.h [hunk 1]
```diff
@@ -234,6 +234,32 @@ class Hl2RxDsp : public QObject {
         return ago < 0 ? 0 : ago;
     }
 
+    // ── Panadapter integrity across a transport gap ───────────────────────
+    //
+    // How many partial FFT frames were thrown away because EP6 packets went
+    // missing part-way through building one. MONOTONIC for the life of this
+    // DSP stage; a reconfigure() builds a new Hl2Spectrum and this count starts
+    // again with it, which is honest -- the frames it counts belonged to a
```
[the remark this answers is not in the document]
Comment (on8st) [reply src/core/backends/hl2/Hl2RxDsp.h]: Correct, and your suggestion taken as written. `40e983f1`.

I verified it the way you did: `grep m_spectrumGapDiscards src/` returns exactly one write in the tree — the `fetch_add` in `onSequenceGap()`. Nothing in `configure()` touches it, and `Hl2Backend` reconfigures the existing `r.dsp` rather than replacing it, so the count outlives the `Hl2Spectrum` it was counting for.

Your framing is the right one and I have kept it: the behaviour is defensible and is unchanged — it is monotonic per stage, which was the first clause and was true. Only the claim was wrong. The comment now says so, and says which event *does* give a fresh count (a receiver torn down and rebuilt) rather than leaving the reader to infer it.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

File: docs/automation-bridge.md [hunk 2]
```diff
@@ -3406,6 +3406,26 @@ other side has.
...
+EP6 blocks at the 1024-point FFT the backend runs, so the two diverge in
+**both** directions: a gap that falls on a frame boundary adds to `droppedPackets` and
+nothing here, while a single lost packet mid-frame costs a frame. A soak test
+that wants to know whether the spectrum it is about to measure is trustworthy
+must threshold this row — `droppedPackets` alone cannot answer it, and a
+non-zero `droppedPackets` with a zero `spectrumGapDiscards0` is a healthy,
+expected state on a lossy link. Both are monotonic within one DSP geometry:
+changing sample rate or receiver count rebuilds the receive chain and restarts
+`spectrumGapDiscards<n>` from zero, so compare a **delta** across a run rather
+than an absolute against a session. The row is `null` for a receiver between
```
[the remark this answers is not in the document]
Comment (on8st) [reply docs/automation-bridge.md]: Taken, with your wording. This is the half that actually costs someone something, and the failure you name is the reason.

A soak script following that paragraph would switch sample rate to zero the counter before its measurement window, carry every earlier discard into that window, and then fail a healthy run — and it would look like a panadapter fault rather than like a test that zeroed nothing. I had written an instruction that breaks the exact thing the row exists to support.

The doc now states the mechanism correctly and adds the negative in as many words: **do not use a geometry change as a way to zero the row.** Your "compare a delta" advice is kept, since it was right and is what a script should do instead.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

File: src/core/backends/anan/P2Client.cpp [hunk 3]
```diff
@@ -285,6 +285,11 @@ void P2Client::onReadyRead()
         if (m_expectedSeq[slot] && frame->seq != *m_expectedSeq[slot]) {
             ++m_drops;
             emit dropsUpdated(m_drops);
+            // The DSP-facing half of the same fact, and BEFORE the decode
+            // below, so a consumer discards its pre-gap state ahead of the
+            // post-gap samples rather than after them. See ddcSequenceGap's
+            // note in the header.
+            emit ddcSequenceGap(*ddcIndex);
```
[the remark this answers is not in the document]
Comment (on8st) [reply src/core/backends/anan/P2Client.cpp]: Taken — I applied the rule rather than documenting the asymmetry, because the asymmetry defeats the row's stated purpose.

You put the cost precisely: this feeds a row the docs tell scripts to **threshold**, and a threshold that counts a sequence restart as loss on one backend and not on the other is not a cross-backend threshold at all. Documenting it would have left every consumer to special-case ANAN.

`P2Client` now guards the new signal with the same `step < 0x80000000u` test `MetisClient` uses. **The `++m_drops` above it is deliberately untouched** — that bare inequality is pre-existing ANAN behaviour, and changing what `m_drops` means is a different question from the one this PR asked. Only the signal this commit introduced is guarded, so nothing that was already being reported changes.

This was also flagged independently before you saw it — it was in my own notes as an inherited inconsistency I had decided to leave alone. You are right that "inherited" stopped being a defence the moment I built a documented, script-facing row on top of it.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

File: tests/spectrum_sequence_gap_test.cpp [hunk 4]
```diff
@@ -0,0 +1,454 @@
...
+        // datagram, or a consumer discards the post-gap samples it had just
+        // been handed.
+        check(order == "BBBGB",
+              "the gap is emitted BEFORE the IQ block of the datagram that revealed it");
+
+        // A BACKWARD jump is the radio restarting its counter, not loss --
+        // MetisClient's existing `gap < 0x80000000u` rule. It must not reach the
+        // DSP either, or every stream restart would throw away a good frame.
+        MetisClientTestAccess::feedDatagram(c, makeEp6(2));
+        check(gaps.size() == 1, "a backward sequence jump is not a gap");
```
[the remark this answers is not in the document]
Comment (on8st) [reply tests/spectrum_sequence_gap_test.cpp]: Both halves accepted as accurate, and I am taking neither in this PR. Here is the reasoning, so you can push back if you disagree.

**"Delete either `connect` and this whole file still passes."** That is true and it is the sharpest thing on this review — the file proves the machinery works and does not prove it is *wired*. I have not closed it, for the reason you gave yourself: the backend `connect`s need more scaffolding than they are worth here.

**The ANAN wire edge is further out of reach than it looks.** `MetisClientTestAccess` works because `MetisClient` already has a separable `handleDatagram(std::span<const std::uint8_t>)` that the friend struct can feed recorded bytes to. `P2Client` has no equivalent: the sequence check lives inline in `onReadyRead()`, which reads the socket directly. A socket-free ANAN test therefore needs `onReadyRead` split into a socket half and a datagram half first — a refactor of the live ANAN receive path, which I do not think belongs in a review round on a panadapter fix. It is a good change and I would rather propose it on its own.

So the honest position, and it is now the one the PR states: **the HL2 wire edge is covered and the ANAN wire edge is not.** Your point that this makes the title's "in both raw-IQ backends" true of the code but not of the evidence is fair, and I would rather a reader knows that than infers symmetry from the title.

One thing your comment did change in the code: the property section 5 pins for `MetisClient` — *"a backward sequence jump is not a gap"* — was not merely untested on ANAN, it was **not true** there. `P2Client` was discarding a frame on a sequence restart. That is fixed in `40e983f1`, so the two backends now agree on the classification even though only one of them has a test saying so.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

**Review by rfoust — APPROVED** [review]
The PR now addresses partial FFT contamination at accepted transport discontinuities in both raw-IQ backends. There is no linked issue; reviewed against the stated fix. The initial independent review reproduced the remaining rewind/duplicate defect on HL2, and the operator authorized repair, approval, and squash merge.

## Scope

| File/group | Claimed in description? | Verdict |
|---|---|---|
| MetisClient and P2Client | Yes | Continuity notification before IQ; ANAN private datagram seam enables socket-free evidence |
| Both spectrum and receive DSP classes | Yes | Reset nonempty windows and count discards |
| HL2/ANAN backend wiring | Yes | Existing I/O-thread delivery pattern; per-receiver routing |
| HL2 health rows and automation documentation | Yes | Additive per-receiver diagnostic with explicit lifetime and continuity semantics |
| Regression test and CMake registration | Yes | Focused socket-free coverage; frozen PR execution gate unchanged |

Everything in the diff is explained by the fix. No unrelated UI/default/persistence/TX changes. The script-facing `spectrumGapDiscards<n>` addition is explicitly documented; operator approval includes this reviewed scope.

## Blockers

None remaining. The backward-jump advice from the earlier review conflated loss with continuity: the packets are still accepted, so a rewind/duplicate must invalidate the pending window even if it adds no forward packet loss. Both transports now notify before accepting those samples. Legacy loss accounting is unchanged. Header and bridge documentation correctly state that counters survive reconfiguration.

## Verification

- Final head: `a1ec80fa213750136404266ee333be8966cb129c`, including main `aea9f6dc4c62c6cc8ad84159e7050ea6a1e728f8`. Main subsequently advanced to `5c146c68b52581d7ccbf9113fe0b4f56b1bf270e`; its combined tree `6d8b2a40ebb76cf41c00764edb4dbff57cee67d1` was rebuilt locally and all three tests passed again. Hosted CI covers the published head on the earlier base; the latest-base evidence is local.
- Local ARM64 RelWithDebInfo build, RADE enabled: `spectrum_sequence_gap_test`, `hl2_spectrum_test`, `anan_p2_protocol_test`; all three pass. No RNNoise x86 sources in the build graph.
- Restoring the old forward-only notification guards produces 29 assertion failures; restoring the repair returns the regression to green. Earlier reset-handler no-op mutations produced four failures.
- Tested both transports through real DSP stages: forward loss, rewind, duplicate, ordinary sequence progression, wraparound, loss across wraparound, half-range jump, exact clean-frame equality and gap-before-IQ ordering. Also tested rolling accumulators, empty windows, ANAN per-DDC isolation and rejected datagrams not advancing sequence state.
- All four architecture gates were run at base and head, then rerun at latest main and the tested combined tree using the updated checkers; both comparisons had identical per-file findings. Engine boundary (tracked legacy warnings only), test registration, frozen CI test-gate and touchpoint manifest checks pass.
- Fresh hosted CI passed before this approval. This does not imply the full suite ran: the new regression is not in the frozen PR execution gate.

Non-blocking limitation: backend constructor wiring is source-reviewed; the test wires the real clients and DSPs itself. The former ANAN-ingest coverage gap is closed. The remaining constructor-wiring coverage is explicitly deferred in the PR description and accepted as non-blocking. No GUI or hardware session was run: SimBackend does not exercise these ingest paths. No sockets, synthetic peers, or TX were used.
